search-service/src/main/kotlin/com/egm/stellio/search/entity/service/EntityAttributeService.kt

```diff
@@ -329,79 +298,122 @@ class EntityAttributeService(
-    ): Either<APIException, Unit> =
-        either {
-            logger.debug("Deleting attribute {} from entity {} (all: {})", attributeName, entityId, deleteAll)
-            if (deleteAll) {
-                attributeInstanceService.deleteAllInstancesOfAttribute(entityId, attributeName).bind()
-                deleteAllInstances(entityId, attributeName).bind()
-            } else {
-                attributeInstanceService.deleteInstancesOfAttribute(entityId, attributeName, datasetId).bind()
-                deleteSpecificInstance(entityId, attributeName, datasetId).bind()
-            }
-        }
+    ): Either<APIException, Unit> = either {
+        logger.debug("Permanently deleting attribute {} from entity {} (all: {})", attributeName, entityId, deleteAll)
+        val attributesToDelete =
+            if (deleteAll)
+                getForEntity(entityId, setOf(attributeName), emptySet(), false)
+            else
+                listOf(getForEntityAndAttribute(entityId, attributeName, datasetId).bind())
 
-    @Transactional
-    suspend fun deleteSpecificInstance(
-        entityId: URI,
-        attributeName: String,
-        datasetId: URI?
-    ): Either<APIException, Unit> =
         databaseClient.sql(
             """
             DELETE FROM temporal_entity_attribute
-            WHERE entity_id = :entity_id
-            ${datasetId.toDatasetIdFilter()}
-            AND attribute_name = :attribute_name
+            WHERE id IN(:uuids)
             """.trimIndent()
         )
-            .bind("entity_id", entityId)
-            .bind("attribute_name", attributeName)
-            .let {
-                if (datasetId != null) it.bind("dataset_id", datasetId)
-                else it
-            }
+            .bind("uuids", attributesToDelete.map { it.id })
             .execute()
 
+        if (deleteAll)
+            attributeInstanceService.deleteAllInstancesOfAttribute(entityId, attributeName).bind()
+        else
+            attributeInstanceService.deleteInstancesOfAttribute(entityId, attributeName, datasetId).bind()
+    }
+
     @Transactional
-    suspend fun deleteAllInstances(
+    suspend fun permanentlyDeleteAttributes(
         entityId: URI,
-        attributeName: String
-    ): Either<APIException, Unit> =
-        databaseClient.sql(
+    ): Either<APIException, Unit> = either {
+        logger.debug("Permanently deleting all attributes from entity {}", entityId)
+
+        val deletedTeas = databaseClient.sql(
             """
             DELETE FROM temporal_entity_attribute
             WHERE entity_id = :entity_id
-            AND attribute_name = :attribute_name
+            RETURNING id
             """.trimIndent()
         )
             .bind("entity_id", entityId)
-            .bind("attribute_name", attributeName)
-            .execute()
+            .allToMappedList { toUuid(it["id"]) }
+
+        attributeInstanceService.deleteInstancesOfEntity(deletedTeas).bind()
```

Comment:
?

Reply:
oh, the second delete is for the attribute_instance table.
i got confused because we filter on a column named "temporal_entity_attribute". Can we easily rename her temporal_entity_attribute_id?

Reply:
we could, but not in the PR.